# __CRITICAL DATA CLEANING AND MODELING PROCESS__

## Step 1: Load and combine the datasets

In [1]:
import pandas as pd
import os

# Directory containing the CSV files
dir_path = r'C:\Users\roen\Desktop\Google Project\Bike Project\original_csv_db\data'

# List of CSV files to combine
csv_files = [
    f'202201-divvy-tripdata.csv',
    f'202202-divvy-tripdata.csv',
    f'202203-divvy-tripdata.csv',
    f'202204-divvy-tripdata.csv',
    f'202205-divvy-tripdata.csv',
    f'202206-divvy-tripdata.csv',
    f'202207-divvy-tripdata.csv',
    f'202208-divvy-tripdata.csv',
    f'202209-divvy-tripdata.csv',
    f'202210-divvy-tripdata.csv',
    f'202211-divvy-tripdata.csv',
    f'202212-divvy-tripdata.csv'
]

# Initialize an empty list to hold dataframes
dfs = []

# Iterate over the list of files and read them into dataframes
for csv_file in csv_files:
    file_path = os.path.join(dir_path, csv_file)
    df = pd.read_csv(file_path)
    dfs.append(df)
    print(f'{csv_file} has {len(df)} rows.')

# Concatenate all dataframes into one
combined_df = pd.concat(dfs, ignore_index=True)

# Save the combined dataframe to a new CSV file
save_path = r'C:\Users\roen\Desktop\Google Project\Bike Project\original_csv_db\combined_data'
combined_csv_path = os.path.join(save_path, 'combined_divvy_tripdata.csv')
combined_df.to_csv(combined_csv_path, index=False)

# Report the row count of the combined dataframe
print(f'Combined dataset has {len(combined_df)} rows.')

# Report the operation is complete
print('Operation combine complete.')

202201-divvy-tripdata.csv has 103770 rows.
202202-divvy-tripdata.csv has 115609 rows.
202203-divvy-tripdata.csv has 284042 rows.
202204-divvy-tripdata.csv has 371249 rows.
202205-divvy-tripdata.csv has 634858 rows.
202206-divvy-tripdata.csv has 769204 rows.
202207-divvy-tripdata.csv has 823488 rows.
202208-divvy-tripdata.csv has 785932 rows.
202209-divvy-tripdata.csv has 701339 rows.
202210-divvy-tripdata.csv has 558685 rows.
202211-divvy-tripdata.csv has 337735 rows.
202212-divvy-tripdata.csv has 181806 rows.
Combined dataset has 5667717 rows.
Operation combine complete.


##### Combined All Raw .csv Files from January to December 2022

## Step 2: Import the Combined Datasets into a DataFrame

In [2]:
# Import necessary libraries
import pandas as pd

# Load the combined dataset
file_path = combined_csv_path
df = pd.read_csv(file_path)

# Display the first few rows
df.head()

ride_id  rideable_type           started_at             ended_at  \
0  C2F7DD78E82EC875  electric_bike  2022-01-13 11:59:47  2022-01-13 12:02:44   
1  A6CF8980A652D272  electric_bike  2022-01-10 08:41:56  2022-01-10 08:46:17   
2  BD0F91DFF741C66D   classic_bike  2022-01-25 04:53:40  2022-01-25 04:58:01   
3  CBB80ED419105406   classic_bike  2022-01-04 00:18:04  2022-01-04 00:33:00   
4  DDC963BFDDA51EEA   classic_bike  2022-01-20 01:31:10  2022-01-20 01:37:12   

              start_station_name start_station_id  \
0       Glenwood Ave & Touhy Ave              525   
1       Glenwood Ave & Touhy Ave              525   
2  Sheffield Ave & Fullerton Ave     TA1306000016   
3       Clark St & Bryn Mawr Ave     KA1504000151   
4    Michigan Ave & Jackson Blvd     TA1309000002   

                end_station_name end_station_id  start_lat  start_lng  \
0           Clark St & Touhy Ave         RP-007  42.012800 -87.665906   
1           Clark St & Touhy Ave         RP-007  42.012763 -87.665967   
2  Greenview Ave & Fullerton Ave   TA1307000001  41.925602 -87.653708   
3      Paulina St & Montrose Ave   TA1309000021  41.983593 -87.669154   
4         State St & Randolph St   TA1305000029  41.877850 -87.624080   

     end_lat    end_lng member_casual  
0  42.012560 -87.674367        casual  
1  42.012560 -87.674367        casual  
2  41.925330 -87.665800        member  
3  41.961507 -87.671387        casual  
4  41.884621 -87.627834        member

## Step 3: Data Cleaning

### 3.1 Understanding the Dataset

In [3]:
# Display basic information about the dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5667717 entries, 0 to 5667716
Data columns (total 13 columns):
 #   Column              Dtype  
---  ------              -----  
 0   ride_id             object 
 1   rideable_type       object 
 2   started_at          object 
 3   ended_at            object 
 4   start_station_name  object 
 5   start_station_id    object 
 6   end_station_name    object 
 7   end_station_id      object 
 8   start_lat           float64
 9   start_lng           float64
 10  end_lat             float64
 11  end_lng             float64
 12  member_casual       object 
dtypes: float64(4), object(9)
memory usage: 562.1+ MB


##### Preliminary Exploration of Data Types for Each Column

### 3.2 Checking for Missing Values

In [4]:
# Check for missing values
missing_values = df.isnull().sum()
missing_values

ride_id                    0
rideable_type              0
started_at                 0
ended_at                   0
start_station_name    833064
start_station_id      833064
end_station_name      892742
end_station_id        892742
start_lat                  0
start_lng                  0
end_lat                 5858
end_lng                 5858
member_casual              0
dtype: int64

### 3.3 Handling Missing Values

In [5]:
# Count rows before removing missing values
total_rows_before = len(df)
print(f'Total rows before removing missing values: {total_rows_before}')

# Drop rows with missing values
df.dropna(inplace=True)

# Count rows after removing missing values
total_rows_after = len(df)
print(f'Total rows after removing missing values: {total_rows_after}')

# Report the number of rows deleted and the percentage
total_deleted = total_rows_before - total_rows_after
percentage_deleted = (total_deleted / total_rows_before) * 100
print(f'Total rows deleted: {total_deleted}')
print(f'Percentage of rows deleted: {percentage_deleted:.2f}%')

Total rows before removing missing values: 5667717
Total rows after removing missing values: 4369360
Total rows deleted: 1298357
Percentage of rows deleted: 22.91%


##### Missing Values Eliminated: Removed 1,298,354 of 5,667,717 Rows (22.91%)

### 3.4 Checking Duplicated Values

In [6]:
# Count rows before removing duplicates
total_rows_before_duplicates = len(df)

# Remove duplicates
df.drop_duplicates(inplace=True)

# Count rows after removing duplicates
total_rows_after_duplicates = len(df)

# Report the number of duplicate rows removed and the percentage
total_duplicates_removed = total_rows_before_duplicates - total_rows_after_duplicates
percentage_duplicates_removed = (total_duplicates_removed / total_rows_before_duplicates) * 100
print(f'Total duplicate rows removed: {total_duplicates_removed}')
print(f'Percentage of duplicate rows removed: {percentage_duplicates_removed:.2f}%')

Total duplicate rows removed: 0
Percentage of duplicate rows removed: 0.00%


### 3.5 Data Modeling

In [7]:
# Convert 'started_at' and 'ended_at' to datetime
df['started_at'] = pd.to_datetime(df['started_at'])
df['ended_at'] = pd.to_datetime(df['ended_at'])

# Create 'ride_length' column as the difference in seconds
df['ride_length'] = (df['ended_at'] - df['started_at']).dt.total_seconds()

# Ensure 'ride_length' is numeric
df['ride_length'] = pd.to_numeric(df['ride_length'], errors='coerce')

# Check if 'ride_id' is unique
is_unique = df['ride_id'].is_unique
print(f'Is ride_id unique: {is_unique}')

# Display the first few rows to check the new column
df.head()

Is ride_id unique: True


ride_id  rideable_type          started_at            ended_at  \
0  C2F7DD78E82EC875  electric_bike 2022-01-13 11:59:47 2022-01-13 12:02:44   
1  A6CF8980A652D272  electric_bike 2022-01-10 08:41:56 2022-01-10 08:46:17   
2  BD0F91DFF741C66D   classic_bike 2022-01-25 04:53:40 2022-01-25 04:58:01   
3  CBB80ED419105406   classic_bike 2022-01-04 00:18:04 2022-01-04 00:33:00   
4  DDC963BFDDA51EEA   classic_bike 2022-01-20 01:31:10 2022-01-20 01:37:12   

              start_station_name start_station_id  \
0       Glenwood Ave & Touhy Ave              525   
1       Glenwood Ave & Touhy Ave              525   
2  Sheffield Ave & Fullerton Ave     TA1306000016   
3       Clark St & Bryn Mawr Ave     KA1504000151   
4    Michigan Ave & Jackson Blvd     TA1309000002   

                end_station_name end_station_id  start_lat  start_lng  \
0           Clark St & Touhy Ave         RP-007  42.012800 -87.665906   
1           Clark St & Touhy Ave         RP-007  42.012763 -87.665967   
2  Greenview Ave & Fullerton Ave   TA1307000001  41.925602 -87.653708   
3      Paulina St & Montrose Ave   TA1309000021  41.983593 -87.669154   
4         State St & Randolph St   TA1305000029  41.877850 -87.624080   

     end_lat    end_lng member_casual  ride_length  
0  42.012560 -87.674367        casual        177.0  
1  42.012560 -87.674367        casual        261.0  
2  41.925330 -87.665800        member        261.0  
3  41.961507 -87.671387        casual        896.0  
4  41.884621 -87.627834        member        362.0

##### Create Critical Column 'ride_length' to Understand User Ride Time and Ensure 'ride_id' Contains Unique Values

### 3.6 Filter Out Anomalous Short Rides

In [8]:
# Count rows before removing short rides
total_rows_before_short_rides = len(df)

# Remove rides with ride_length less than 10 seconds
df = df[df['ride_length'] > 60]

# Count rows after removing short rides
total_rows_after_short_rides = len(df)

# Report the number of short rides removed and the percentage
total_deleted_short_rides = total_rows_before_short_rides - total_rows_after_short_rides
percentage_deleted_short_rides = (total_deleted_short_rides / total_rows_before_short_rides) * 100
print(f'Total short rides (less than 10 seconds) removed: {total_deleted_short_rides}')
print(f'Percentage of short rides removed: {percentage_deleted_short_rides:.2f}%')

Total short rides (less than 10 seconds) removed: 77399
Percentage of short rides removed: 1.77%


#### __Ensure 'ride_length' is Greater Than 60 Seconds to Eliminate Short Rides__

##### I assumed rides shorter than 60 seconds might indicate users who changed their minds or unintentionally started the ride. By removing these short rides, 1.77% of the dataset, totaling 77,399 rows, was eliminated. This step improves the reliability of the analysis

### Step 4: Save the Cleaned Dataset

In [19]:
# Define the save path and file name
save_path = r'C:\Users\roen\Desktop\Google Project\Bike Project\original_csv_db\cleaned_data'
file_name = 'cleaned_divvy_tripdata_clean.csv'

# Combine the save path and file name
full_save_path = f'{save_path}\{file_name}'

# Save the cleaned dataset to a new CSV file
df.to_csv(full_save_path, index=False)

# Report the row count of the cleaned dataframe
cleaned_percentage = abs((len(df) / len(combined_df) - 1) * 100)
print(f'The cleaned dataset contains a total of {len(df)} rows.')
print(f'The combined dataset contains {len(combined_df)} rows.')
print(f'Total percentage of data cleaned: {cleaned_percentage:.2f}%')

# Report the operation is complete
print('Data cleaning and saving complete.')

The cleaned dataset contains a total of 4291961 rows.
The combined dataset contains 5667717 rows.
Total percentage of data cleaned: 24.27%
Data cleaning and saving complete.
